README.md
# aandd_ekew_driver_py

## install python module "serial"
```.sh
sudo rosdep init
rosdep update
rosdep install --from-paths aandd_ekew_driver_py -y --ignore-src
```

## prepare for interfaces (option)
this node uses weight_scale_interfaces
```.sh
cd ~/dev_ws/src
git clone https://github.com/TechMagicKK/weight_scale_interfaces.git
cd ~/dev_ws
colcon build --cmake-clean-first --symlink-install --packages-select weight_scale_interfaces
. install/local_setup.zsh
```

## rs-232c device setting
```.sh
cd ~/dev_ws/aandd_ekew_driver_py/launch
edit bringup.launch.py for change params (port, baudrate, rate)
```

## build
```.sh
cd ~/dev_ws
colcon build --cmake-clean-first --symlink-install --packages-select aandd_ekew_driver_py
. install/local_setup.zsh
```

## launch node
```.sh
ros2 launch aandd_ekew_driver_py bringup.launch.py
```

## launch node with fake device for test
```.sh
ros2 launch aandd_ekew_driver_py bringup.launch.py use_fake:=True
```

## change publish rate to 1.2[Hz]
```.sh
ros2 param set /aandd_ekew_node rate 1.2
```



CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package aandd_ekew_driver_py
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.2 (2023-09-19)
------------------
* Added author info
* Contributors: Jiaqing Lin

0.0.1 (2023-09-05)
------------------
* relesase
* Contributors: Jiaqing Lin


aandd_ekew_driver/aandd_ekew_test.py
import time
import rclpy
from rclpy.node import Node
from rclpy.action import ActionClient
from rclpy.executors import MultiThreadedExecutor
from action_msgs.msg import GoalStatus
from weight_scale_interfaces.msg import Weight
from weight_scale_interfaces.action import SetZero, GetWeight

class WeightAndScaleTestNode(Node):
    def __init__(self):
        super().__init__('aandd_ekew_test')
        self._weight_sub = self.create_subscription(Weight, '/aandd_ekew_node/weight', self.weight_callback, 10)
        self._set_zero_goal_handle = None
        self._set_zero_action = ActionClient(self, SetZero, '/aandd_ekew_node/set_zero')
        self._get_weight_goal_handle = None
        self._get_weight_action = ActionClient(self, GetWeight, '/aandd_ekew_node/get_weight')

    # ----------------------------------------------------------------------------------------------------------------------------
    async def weight_callback(self, msg):
        #self.get_logger().info(f'published_weight({msg.stamp.sec}.{msg.stamp.nanosec:09}, {msg.stable}, {msg.overload}, {msg.weight:.2f}[{msg.unit}])')
        pass

    # ----------------------------------------------------------------------------------------------------------------------------
    def cancel_set_zero(self):
        if self._set_zero_goal_handle is not None:
            future = self._set_zero_goal_handle.cancel_goal_async()
            response = future.result()

    async def set_zero_feedback(self, feedback_msg):
        feedback = feedback_msg.feedback
        msg = feedback.weight
        self.get_logger().info(f'set_zero_feedback: {msg.stamp.sec}.{msg.stamp.nanosec:09}, {msg.stable}, {msg.overload}, {msg.weight:.2f}[{msg.unit}]')

    def set_zero(self, timeout_sec):
        self._set_zero_action.wait_for_server()
        goal = SetZero.Goal()
        goal.timeout = timeout_sec
        send_goal_future = self._set_zero_action.send_goal_async(
            goal,
            feedback_callback=self.set_zero_feedback)
        rclpy.spin_until_future_complete(self, send_goal_future, timeout_sec=20.0)
        self._set_zero_goal_handle = send_goal_future.result()
        if (self._set_zero_goal_handle is None) or (not self._set_zero_goal_handle.accepted):
            self.get_logger().info('exception: set_zero: goal')
            raise Exception('set_zero: goal')
        get_result_future = self._set_zero_goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, get_result_future, timeout_sec=20.0)
        result = get_result_future.result()
        if result is None:
            self.get_logger().info('exception: set_zero: result')
            raise Exception('set_zero: result')
        self._set_zero_goal_handle = None
        return result.result

    # ----------------------------------------------------------------------------------------------------------------------------
    def cancel_get_weight(self):
        if self._get_weight_goal_handle is not None:
            future = self._get_weight_goal_handle.cancel_goal_async()
            response = future.result()

    async def get_weight_feedback(self, feedback_msg):
        feedback = feedback_msg.feedback
        msg = feedback.weight
        self.get_logger().info(f'get_weight_feedback: {msg.stamp.sec}.{msg.stamp.nanosec:09}, {msg.stable}, {msg.overload}, {msg.weight:.2f}[{msg.unit}]')

    def get_weight(self, timeout_sec):
        self._get_weight_action.wait_for_server()
        goal = GetWeight.Goal()
        goal.timeout = timeout_sec
        send_goal_future = self._get_weight_action.send_goal_async(
            goal,
            feedback_callback=self.get_weight_feedback)
        rclpy.spin_until_future_complete(self, send_goal_future, timeout_sec=20.0)
        self._get_weight_goal_handle = send_goal_future.result()
        if (self._get_weight_goal_handle is None) or (not self._get_weight_goal_handle.accepted):
            self.get_logger().info('exception: get_weight: goal')
            raise Exception('get_weight: goal')
        get_result_future = self._get_weight_goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, get_result_future, timeout_sec=20.0)
        result = get_result_future.result()
        if result is None:
            self.get_logger().info('exception: get_weight: result')
            raise Exception('get_weight: result')
        self._get_weight_goal_handle = None
        return result.result

# ----------------------------------------------------------------------------------------------------------------------------
def spin_sleep(executor, sec):
    future = executor._executor.submit(lambda: time.sleep(sec))
    executor.spin_until_future_complete(future)

def main(args=None):
    rclpy.init(args=args)
    executor = rclpy.executors.MultiThreadedExecutor()
    node = WeightAndScaleTestNode()
    executor.add_node(node)
    try:
        res = node.set_zero(timeout_sec=1.0)
        node.get_logger().info(f'set_zero: success={res.success}, message={res.message}')
        while rclpy.ok():
            res = node.get_weight(timeout_sec=1.0)
            msg = res.weight
            node.get_logger().info(f'get_weight: {msg.stamp.sec}.{msg.stamp.nanosec:09}, {msg.stable}, {msg.overload}, {msg.weight:.2f}[{msg.unit}], {res.success}, {res.message}')
            spin_sleep(executor, 1)
    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        node.destroy_node()

if __name__ == '__main__':
    main()


aandd_ekew_driver/aandd_ekew_driver_test.py
import signal
import time
import threading
import rclpy
from rclpy.node import Node
from rclpy.signals import SignalHandlerOptions 
from rclpy.action import ActionClient
from rclpy.executors import MultiThreadedExecutor, SingleThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup, MutuallyExclusiveCallbackGroup

from action_msgs.msg import GoalStatus

from weight_scale_interfaces.msg import Weight
from weight_scale_interfaces.action import SetZero
from weight_scale_interfaces.action import GetWeight

class WeightAndScaleTestNode(Node):
    def __init__(self):
        super().__init__('aandd_ekew_driver_test')
        self._weight_sub = self.create_subscription(Weight, '/aandd_ekew_driver_node/weight', self.weight_callback, 10)
        self._set_zero_goal_handle = None
        self._set_zero_action = ActionClient(self, SetZero, '/aandd_ekew_driver_node/set_zero')
        self._get_weight_goal_handle = None
        self._get_weight_action = ActionClient(self, GetWeight, '/aandd_ekew_driver_node/get_weight')

    # ----------------------------------------------------------------------------------------------------------------------------
    async def weight_callback(self, msg):
        self.get_logger().info(f'published_weight({msg.stamp.sec}.{msg.stamp.nanosec:*<9}, {msg.stable}, {msg.weight}[{msg.unit}])')

    # ----------------------------------------------------------------------------------------------------------------------------
    def cancel_set_zero(self):
        if self._set_zero_goal_handle is not None:
            future = self._set_zero_goal_handle.cancel_goal_async()
            response = future.result()

    async def set_zero_feedback(self, feedback_msg):
        feedback = feedback_msg.feedback
        msg = feedback.weight
        self.get_logger().info(f'set_zero_feedback: {msg.stamp.sec}.{msg.stamp.nanosec:*<9}, {msg.stable}, {msg.weight:.2f}[{msg.unit}]')

    def set_zero(self, timeout_sec):
        self._set_zero_action.wait_for_server()
        goal = SetZero.Goal()
        goal.timeout = timeout_sec
        send_goal_future = self._set_zero_action.send_goal_async(
            goal,
            feedback_callback=self.set_zero_feedback)
        rclpy.spin_until_future_complete(self, send_goal_future, timeout_sec=10.0)
        self._set_zero_goal_handle = send_goal_future.result()
        if (self._set_zero_goal_handle is None) or (not self._set_zero_goal_handle.accepted):
            self.get_logger().info('exception: set_zero: goal')
            raise Exception('set_zero: goal')
        get_result_future = self._set_zero_goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, get_result_future, timeout_sec=10.0)
        result = get_result_future.result()
        if result is None:
            self.get_logger().info('exception: set_zero: result')
            raise Exception('set_zero: result')
        self._set_zero_goal_handle = None
        return result.result

    # ----------------------------------------------------------------------------------------------------------------------------
    def cancel_get_weight(self):
        if self._get_weight_goal_handle is not None:
            future = self._get_weight_goal_handle.cancel_goal_async()
            response = future.result()

    async def get_weight_feedback(self, feedback_msg):
        feedback = feedback_msg.feedback
        msg = feedback.weight
        self.get_logger().info(f'get_weight_feedback: {msg.stamp.sec}.{msg.stamp.nanosec:*<9}, {msg.stable}, {msg.weight:.2f}[{msg.unit}]')

    def get_weight(self, timeout_sec):
        self._get_weight_action.wait_for_server()
        goal = GetWeight.Goal()
        goal.timeout = timeout_sec
        send_goal_future = self._get_weight_action.send_goal_async(
            goal,
            feedback_callback=self.get_weight_feedback)
        rclpy.spin_until_future_complete(self, send_goal_future, timeout_sec=10.0)
        self._get_weight_goal_handle = send_goal_future.result()
        if (self._get_weight_goal_handle is None) or (not self._get_weight_goal_handle.accepted):
            self.get_logger().info('exception: get_weight: goal')
            raise Exception('get_weight: goal')
        get_result_future = self._get_weight_goal_handle.get_result_async()
        rclpy.spin_until_future_complete(self, get_result_future, timeout_sec=10.0)
        result = get_result_future.result()
        if result is None:
            self.get_logger().info('exception: get_weight: result')
            raise Exception('get_weight: result')
        self._get_weight_goal_handle = None
        return result.result

# ----------------------------------------------------------------------------------------------------------------------------
def main(args=None):
    rclpy.init(args=args, signal_handler_options=SignalHandlerOptions.NO)
    node = WeightAndScaleTestNode()
    thread = threading.Thread(target=rclpy.spin, args=(node,), daemon=True)
    thread.start()
    try:
        while rclpy.ok():
            res = node.set_zero(timeout_sec=1.0)
            node.get_logger().info(f'set_zero: success={res.success}, message={res.message}')
            time.sleep(5)
            res = node.get_weight(timeout_sec=1.0)
            msg = res.weight
            node.get_logger().info(f'get_weight: {msg.stamp.sec}.{msg.stamp.nanosec:*<9}, {msg.stable}, {msg.weight:.2f}[{msg.unit}], {res.success}, {res.message}')
            time.sleep(5)
    except Exception as e:
        node.get_logger().info(f'exception: {e}')
        rclpy.shutdown()
        thread.join()

if __name__ == '__main__':
    main()


launch/bringup.launch.py
import launch
import launch.actions
import launch.events

import launch_ros
import launch_ros.actions
import launch_ros.events

from launch import LaunchDescription
from launch_ros.actions import LifecycleNode
from launch_ros.actions import Node

import lifecycle_msgs.msg


def generate_launch_description():
    ld = launch.LaunchDescription()

    port = launch.substitutions.LaunchConfiguration('port', default='/dev/ttyUSB0')
    baudrate = launch.substitutions.LaunchConfiguration('baudrate', default=9600)
    connect_timeout = launch.substitutions.LaunchConfiguration('connect_timeout', default=1.0)
    rate = launch.substitutions.LaunchConfiguration('rate', default=5.0)
    use_fake = launch.substitutions.LaunchConfiguration('use_fake', default=False)

    node = launch_ros.actions.LifecycleNode(
        name='aandd_ekew_node', namespace='',
        package='aandd_ekew_driver_py', executable='aandd_ekew_node', output='screen',
        parameters=[{'port':port, 'baudrate':baudrate, 'connect_timeout':connect_timeout, 'rate':rate, 'use_fake':use_fake}],
    )

    to_inactive = launch.actions.EmitEvent(
        event=launch_ros.events.lifecycle.ChangeState(
            lifecycle_node_matcher=launch.events.matches_action(node),
            transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
        )
    )
    
    to_active = launch.actions.EmitEvent(
        event=launch_ros.events.lifecycle.ChangeState(
            lifecycle_node_matcher=launch.events.matches_action(node),
            transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
        )
    )
    
    from_unconfigured_to_inactive = launch.actions.RegisterEventHandler(
        launch_ros.event_handlers.OnStateTransition(
            target_lifecycle_node=node, 
            goal_state='unconfigured',
            entities=[
                launch.actions.LogInfo(msg="<< Unconfigured >>"),
                launch.actions.EmitEvent(event=launch_ros.events.lifecycle.ChangeState(
                    lifecycle_node_matcher=launch.events.matches_action(node),
                    transition_id=lifecycle_msgs.msg.Transition.TRANSITION_CONFIGURE,
                )),
            ],
        )
    )

    from_inactive_to_active = launch.actions.RegisterEventHandler(
        launch_ros.event_handlers.OnStateTransition(
            target_lifecycle_node=node, 
            start_state = 'configuring',
            goal_state='inactive',
            entities=[
                launch.actions.LogInfo(msg="<< Inactive >>"),
                launch.actions.EmitEvent(event=launch_ros.events.lifecycle.ChangeState(
                    lifecycle_node_matcher=launch.events.matches_action(node),
                    transition_id=lifecycle_msgs.msg.Transition.TRANSITION_ACTIVATE,
                )),
            ],
        )
    )

    from_active_to_inactive = launch.actions.RegisterEventHandler(
        launch_ros.event_handlers.OnStateTransition(
            target_lifecycle_node=node, 
            goal_state='active',
            entities=[
                launch.actions.LogInfo(msg="<< Active >>"),
                launch.actions.EmitEvent(event=launch_ros.events.lifecycle.ChangeState(
                    lifecycle_node_matcher=launch.events.matches_action(node),
                    transition_id=lifecycle_msgs.msg.Transition.TRANSITION_DEACTIVATE,
                )),
            ],
        )
    )

    from_inactive_to_finalized = launch.actions.RegisterEventHandler(
        launch_ros.event_handlers.OnStateTransition(
            target_lifecycle_node=node, 
            start_state = 'deactivating',
            goal_state='inactive',
            entities=[
                launch.actions.LogInfo(msg="<< Inactive >>"),
                launch.actions.EmitEvent(event=launch_ros.events.lifecycle.ChangeState(
                    lifecycle_node_matcher=launch.events.matches_action(node),
                    transition_id=lifecycle_msgs.msg.Transition.TRANSITION_INACTIVE_SHUTDOWN,
                )),
            ],
        )
    )
    
    from_finalized_to_exit = launch.actions.RegisterEventHandler(
        launch_ros.event_handlers.OnStateTransition(
            target_lifecycle_node=node, goal_state='finalized',
            entities=[
                launch.actions.LogInfo(msg="<< Finalized >>"),
                launch.actions.EmitEvent(event=launch.events.Shutdown()),
            ],
        )
    )
    
    #ld.add_action(from_unconfigured_to_inactive)
    ld.add_action(from_inactive_to_active)
    #ld.add_action(from_active_to_inactive)
    #ld.add_action(from_inactive_to_finalized)
    ld.add_action(from_finalized_to_exit)
    
    ld.add_action(node)
    ld.add_action(to_inactive)
    #ld.add_action(to_active)

    return ld


aandd_ekew_driver_py/aandd_ekew_node.py
import time
import re
from serial import Serial
from typing import Optional

import rclpy
from rclpy.executors import MultiThreadedExecutor
from weight_scale_interfaces.msg import Weight
from aandd_ekew_driver_py.weight_scale_node import WeightScaleNode, WeightScaleError

#---------------------------------------------------------------------------------------
class EKEWNode(WeightScaleNode):
    def __init__(self, **kwargs):
        super().__init__('aandd_ekew_node', **kwargs)
        self._client: Optional[Serial] = None
        self.declare_parameter('port', '/dev/ttyUSB0')
        self.declare_parameter('baudrate', 9600)
        self.declare_parameter('connect_timeout', 1.0)
        self.declare_parameter('use_fake', False)
        self._port = self.get_parameter('port').value
        self._baudrate = self.get_parameter('baudrate').value
        self._connect_timeout = self.get_parameter('connect_timeout').value
        self._use_fake = self.get_parameter('use_fake').value
        self.get_logger().info(f'params: port={self._port}')
        self.get_logger().info(f'params: baudrate={self._baudrate}[bps]')
        self.get_logger().info(f'params: connect_timeout={self._connect_timeout}')
        self.get_logger().info(f'params: use_fake={self._use_fake}')

    def connect(self):
        if self._use_fake:
            super().connect()
            return
        try:
            self.disconnect()
            self.get_logger().info(f'connect(port={self._port}, baudrate={self._baudrate}, timeout={self._connect_timeout})')
            self._client = Serial(port=self._port, baudrate=self._baudrate, timeout=self._connect_timeout)
        except Exception as e:
            raise WeightScaleError(f'connect: {e}')

    def disconnect(self):
        if self._use_fake:
            super().disconnect()
            return
        if self._client is not None:
            self._client.close()
            self._client = None

    def set_zero(self):
        if self._use_fake:
            super().set_zero()
            return
        try:
            self._client.write(b'Z\r\n')
            line = self._client.readline()
            if line is None:
                raise Exception
            time.sleep(1.0)
        except Exception as e:
            raise WeightScaleError(f'set_zero: communication error')

    def get_weight(self) -> Weight():
        if self._use_fake:
            return super().get_weight()
        try:
            msg = Weight()
            msg.stable = False
            msg.overload = False
            msg.weight = 0.0
            msg.unit = 'kg'
            msg.stamp = self.get_clock().now().to_msg()
            self._client.write(b'Q\r\n')
            line = self._client.readline().strip().decode('utf-8')
            if line is None:
                raise Exception
            m = re.search(r'(ST|QT|US|OL),([ 0-9\.\+\-]+)([ a-zA-Z\%]+)', line)
            if m.group(1) == 'OL':
                # OverLoad
                msg.overload = True
                msg.unit = m.group(3).lstrip()
            elif (m.group(1) == 'ST') or (m.group(1) == 'QT'):
                # STable
                msg.stable = True
                msg.weight = float(m.group(2))
                msg.unit = m.group(3).lstrip()
            else:
                # must be US(UnStable)
                msg.stable = False
                msg.weight = float(m.group(2))
                msg.unit = m.group(3).lstrip()
            return msg
        except Exception as e:
            raise WeightScaleError(f'get_weight: communication error')

#---------------------------------------------------------------------------------------
def main(args=None):
    try:
        rclpy.init(args=args)
        executor = rclpy.executors.MultiThreadedExecutor()
        node = EKEWNode()
        executor.add_node(node)
        executor.spin()
    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        node.manual_shutdown(debug_out=False)
        node.destroy_node()

if __name__ == '__main__':
    main()


aandd_ekew_driver/aandd_ekew_driver_node.py
import time
import sys
import re
import threading
import random
import asyncio
from serial import Serial
from typing import Optional

import rclpy
from rclpy.timer import Timer
from rclpy.time import Time
from rclpy.action import ActionServer, CancelResponse, GoalResponse
from rclpy.callback_groups import ReentrantCallbackGroup, MutuallyExclusiveCallbackGroup
from rclpy.executors import MultiThreadedExecutor, SingleThreadedExecutor
from rclpy.lifecycle import Node
from rclpy.lifecycle import Publisher
from rclpy.lifecycle import State
from rclpy.lifecycle import TransitionCallbackReturn
from rclpy.qos import QoSProfile, QoSReliabilityPolicy, QoSHistoryPolicy, QoSDurabilityPolicy, QoSPresetProfiles

from weight_scale_interfaces.msg import Weight
from weight_scale_interfaces.action import SetZero
from weight_scale_interfaces.action import GetWeight

#---------------------------------------------------------------------------------------
class EKEWError(Exception):
    pass

class EKEW(object):
    def __init__(self, node):
        self._node = node
        self._client: Optional[Serial] = None
        self._lock = threading.Lock()

    def connect(self, device, baudrate, timeout):
        self.disconnect()
        try:
            with self._lock:
                self._node.get_logger().info(f'Serial(port={device}, baudrate={baudrate}, timeout={timeout})')
                self._client = Serial(port=device, baudrate=baudrate, timeout=timeout)
        except Exception as e:
            raise EKEWError(f'connect: {e}')

    def disconnect(self):
        if self._client is not None:
            with self._lock:
                self._client.close()
                self._client = None
        return True

    def set_zero(self):
        try:
            with self._lock:
                command = 'Z'
                self._client.write(f'{command}\r\n'.encode('utf-8'))
                line = self._client.readline().strip().decode('utf-8')
                if line is not command:
                    raise Exception
        except Exception as e:
            raise EKEWError(f'set_zero: {e}')

    def get_weight(self) -> Weight():
        try:
            msg = Weight()
            msg.stable = False
            msg.overload = False
            msg.weight = 0.0
            msg.unit = 'kg'
            with self._lock:
                msg.stamp = self._node.get_clock().now().to_msg()
                command = 'Q'
                self._client.write(f'{command}\r\n'.encode('utf-8'))
                line = self._client.readline().strip().decode('utf-8')
                m = re.search(r'(ST|QT|US|OL),([ 0-9\.\+\-]+)([ a-zA-Z\%]+)', line)
                if m.group(1) == 'OL':
                    # OverLoad
                    msg.overload = True
                    msg.unit = m.group(3).lstrip()
                elif (m.group(1) == 'ST') or (m.group(1) == 'QT'):
                    # STable
                    msg.stable = True
                    msg.weight = float(m.group(2))
                    msg.unit = m.group(3).lstrip()
                else:
                    # must be US(UnStable)
                    msg.stable = False
                    msg.weight = float(m.group(2))
                    msg.unit = m.group(3).lstrip()
                return msg
        except Exception as e:
            raise EKEWError(f'get_weight: {e}')

class EKEWTest(object):
    def __init__(self, node):
        self._node = node
        self._lock = threading.Lock()
        self._weight_str = 'ST,+01234.56  g'

    def connect(self, device, baudrate, timeout):
        pass

    def disconnect(self):
        pass

    def set_zero(self):
        self._weight_str = 'ST,+00000.00  g'

    def get_weight(self) -> Weight():
        try:
            msg = Weight()
            with self._lock:
                msg.stamp = self._node.get_clock().now().to_msg()
                msg.stable = True if random.random() < 0.1 else False
                msg.overload = False
                line = self._weight_str
                m = re.search(r'(ST|QT|US|OL),([ 0-9\.\+\-]+)([ a-zA-Z\%]+)', line)
                msg.weight = float(m.group(3)) + random.random()
                msg.unit = m.group(4).lstrip()
            return msg
        except Exception as e:
            raise EKEWError(f'get_weight: {e}')

#---------------------------------------------------------------------------------------
class WeightAndScaleNode(Node):
    def __init__(self, **kwargs):
        self._ekew: Optional[EKEW] = None
        self._weight_publisher: Optional[Publisher] = None
        self._weight_publish_timer: Optional[Timer] = None
        self._set_zero_action: Optional[ActionServer] = None
        self._get_weight_action: Optional[ActionServer] = None
        super().__init__('aandd_ekew_driver_node', **kwargs)

    def setup_params(self):
        self.declare_parameter('device', '/dev/ttyUSB0')
        self.declare_parameter('baudrate', 9600)
        self.declare_parameter('connect_timeout', 1.0)
        self.declare_parameter('rate', 5.0)
        self._device = self.get_parameter('device').value
        self._baudrate = self.get_parameter('baudrate').value
        self._connect_timeout = self.get_parameter('connect_timeout').value
        self._rate = self.get_parameter('rate').value
        self.get_logger().info(f'params: device={self._device}')
        self.get_logger().info(f'params: baudrate={self._baudrate}[bps]')
        self.get_logger().info(f'params: connect_timeout={self._connect_timeout}')
        self.get_logger().info(f'params: rate={self._rate}[Hz]')

    async def publish_weight(self):
        try:
            if (self._weight_publisher is not None) and self._weight_publisher.is_activated:
                msg = self._ekew.get_weight()
                self.get_logger().info(f'publish_weight({msg.stamp.sec}.{msg.stamp.nanosec:*<9}, {msg.stable}, {msg.overload}, {msg.weight:.2f}[{msg.unit}])')
                self._weight_publisher.publish(msg)
                rate = self.get_parameter('rate').value
                if rate < 0.01:
                    rate = 0.01
                if (self._weight_publish_timer is not None) and (rate != self._rate):
                    self.get_logger().info(f'change rate param from {self._rate:.2f}[Hz] to {rate:.2f}[Hz]')
                    self.destroy_timer(self._weight_publish_timer)
                    self._rate = rate
                    self._weight_publish_timer = self.create_timer(1.0/self._rate, self.publish_weight)
        except Exception as e:
            raise EKEWError(f'publish_weight: {e}')

    #-----------------------------------------------------------------------------
    def on_configure(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('lifecycle:on_configure()')
        #self._ekew = EKEW(self)
        self._ekew = EKEWTest(self)
        self._weight_publisher = self.create_lifecycle_publisher(Weight, "~/weight", 10);
        self.get_logger().info(f'create_timer({1.0/self._rate:.2f})')
        self._weight_publish_timer = self.create_timer(1.0/self._rate, self.publish_weight)
        self._set_zero_action = ActionServer(
            self,
            SetZero,
            '~/set_zero',
            execute_callback=self.set_zero_execute_callback,
            cancel_callback=self.set_zero_cancel_callback,
            result_timeout = 60
        )
        self._get_weight_action = ActionServer(
            self,
            GetWeight,
            '~/get_weight',
            execute_callback=self.get_weight_execute_callback,
            cancel_callback=self.get_weight_cancel_callback,
            result_timeout = 60
        )
        return TransitionCallbackReturn.SUCCESS

    def on_activate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('lifecycle:on_activate()')
        self.get_logger().info(f'connect(device="{self._device}", baudrate={self._baudrate}[bps], connect_timeout={self._connect_timeout}[s])')
        try:
            self._ekew.connect(self._device, self._baudrate, self._connect_timeout)
        except Exception as e:
            self.get_logger().error(f'exception: {e}')
            return TransitionCallbackReturn.FAILURE
        return super().on_activate(state)

    def on_deactivate(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('lifecycle:on_deactivate()')
        if self._ekew is not None:
            self.get_logger().info('disconnect()')
            self._ekew.disconnect()
        return super().on_deactivate(state)

    def on_cleanup(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('lifecycle:on_cleanup()')
        self.manual_shutdown()
        return TransitionCallbackReturn.SUCCESS

    def on_shutdown(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('lifecycle:on_shutdown()')
        self.manual_shutdown()
        return TransitionCallbackReturn.SUCCESS

    def on_error(self, state: State) -> TransitionCallbackReturn:
        self.get_logger().info('lifecycle:on_error()')
        self.manual_shutdown()
        return TransitionCallbackReturn.SUCCESS

    def manual_shutdown(self, debug_out=True): # after exception
        if self._weight_publish_timer is not None:
            if debug_out: self.get_logger().info('cleanup timer')
            self.destroy_timer(self._weight_publish_timer)
            self._weight_publish_timer = None
        if self._weight_publisher is not None:
            if debug_out: self.get_logger().info('destroy weight_publisher')
            self.destroy_publisher(self._weight_publisher)
            self._weight_publisher = None
        if self._set_zero_action is not None:
            if debug_out: self.get_logger().info('destroy set_zero_actions')
            self._set_zero_action.destroy()
            self._set_zero_action = None
        if self._get_weight_action is not None:
            if debug_out: self.get_logger().info('destroy get_weight_actions')
            self._get_weight_action.destroy()
            self._get_weight_action = None
        if self._ekew is not None:
            if debug_out: self.get_logger().info('disconnect()')
            self._ekew.disconnect()
            self._ekew = None

    #-----------------------------------------------------------------------------
    async def set_zero_cancel_callback(self, goal_handle):
        self.get_logger().info('cancel set_zero()')
        return CancelResponse.ACCEPT

    async def set_zero_execute_callback(self, goal_handle):
        goal = goal_handle.request
        result = SetZero.Result()
        feedback = SetZero.Feedback()
        weight = Weight()
        self.get_logger().info(f'set_zero({goal.timeout:5.2f}[s])')

        try:
            timeout = self.get_clock().now() + rclpy.duration.Duration(seconds=goal.timeout)
            while True:
                weight = self._ekew.get_weight()
                if goal_handle.is_cancel_requested:
                    self.get_logger().info('set_zero: canceled')
                    result.success = False
                    result.message = 'canceled'
                    goal_handle.canceled()
                    return result
                if weight.stable:
                    self.get_logger().info('set_zero: stable')
                    self._ekew.set_zero()
                    result.success = True
                    result.message = 'success'
                    goal_handle.succeed()
                    return result
                elif Time.from_msg(weight.stamp) > timeout:
                    self.get_logger().info('set_zero: timeout')
                    result.success = False
                    result.message = 'timeout'
                    goal_handle.succeed()
                    return result
                else:
                    self.get_logger().info('set_zero: feedback')
                    feedback.weight = weight
                    goal_handle.publish_feedback(feedback)
                    #asyncio.sleep(0.1)
                    time.sleep(0.1)
        except EKEWError as e:
            result.success = False
            result.weight = weight
            result.message = 'modbus communication error'
            goal_handle.succeed()
            return result

    #-----------------------------------------------------------------------------
    async def get_weight_cancel_callback(self, goal_handle):
        self.get_logger().info('cancel set_zero()')
        return CancelResponse.ACCEPT

    async def get_weight_execute_callback(self, goal_handle):
        goal = goal_handle.request
        result = GetWeight.Result()
        feedback = GetWeight.Feedback()
        weight = Weight()
        self.get_logger().info(f'get_weight({goal.timeout:5.2f}[s])')

        try:
            timeout = self.get_clock().now() + rclpy.duration.Duration(seconds=goal.timeout)
            while True:
                weight = self._ekew.get_weight()
                if goal_handle.is_cancel_requested:
                    result.weight = weight
                    result.success = False
                    result.message = 'canceled'
                    goal_handle.canceled()
                    return result
                if weight.stable:
                    result.weight = weight
                    result.success = True
                    result.message = 'success'
                    goal_handle.succeed()
                    return result
                elif Time.from_msg(weight.stamp) > timeout:
                    result.weight = weight
                    result.success = False
                    result.message = 'timeout'
                    goal_handle.succeed()
                    return result
                else:
                    feedback.weight = weight
                    goal_handle.publish_feedback(feedback)
                    #asyncio.sleep(0.1)
                    time.sleep(0.1)
        except EKEWError as e:
            result.success = False
            result.weight = weight
            result.message = 'modbus communication error'
            goal_handle.succeed()
            return result

def main(args=None):
    rclpy.init(args=args)
    executor = rclpy.executors.MultiThreadedExecutor()
    node = WeightAndScaleNode()
    executor.add_node(node)
    node.setup_params()
    try:
        executor.spin()
    except (KeyboardInterrupt, rclpy.executors.ExternalShutdownException):
        node.manual_shutdown(debug_out=False)
        node.destroy_node()

if __name__ == '__main__':
    main()
